### PR TITLE
doc: update storage driver reference

### PR DESCRIPTION
**Pull request by ru-fu**
The section in the ZFS docs about growing the pool will be moved out in a separate PR.

### CODE REVIEW

File: doc/reference/storage_dir.md [hunk 1]
```diff
@@ -1,25 +1,38 @@
...
+This driver is quick to set up and allows inspecting the files directly on the disk, which can be convenient for testing.
+However, LXD operations are not optimized for this driver.
 
-## Storage pool configuration
+## `dir` driver in LXD
+
+The `dir` driver in LXD is fully functional and provides the same set of features as other drivers.
+However, it is much slower than all the other drivers because it must unpack images and do instant copies of instances, snapshots and images.
+
+Unless specified differently during creation (with the `source` configuration option), the data is stored in the `/var/snap/lxd/common/lxd/storage-pools/` directory.
```
Comment (ru-fu) [inline doc/reference/storage_dir.md:13]: The location is probably different for non-snap installations?

Comment (tomponline) [reply doc/reference/storage_dir.md:13]: Yes it will be /var/lib/lxd/storage-pools/....

File: doc/reference/storage_btrfs.md [hunk 2]
```diff
@@ -1,6 +1,18 @@
 (storage-btrfs)=
 # Btrfs - `btrfs`
 
+{abbr}`Btrfs (B-tree file system)` is a single-disk file system based on the {abbr}`COW (copy-on-write)` principle.
```
Comment (stgraber) [inline doc/reference/storage_btrfs.md:4]: Maybe replace `single-disk` with `local` as btrfs actually supports built-in multi-disk layouts including RAID.
Not something we should really cover here but calling it single-disk is inaccurate.

File: doc/reference/storage_btrfs.md [hunk 3]
```diff
@@ -1,6 +1,18 @@
...
+COW means that data is stored to a different block after it has been modified instead of overwriting the existing data, reducing the risk of data corruption.
+Unlike other file systems, Btrfs is extent-based, which means that it stores data in contiguous areas of memory.
+
+In addition to basic file system features, Btrfs offers RAID and volume management, pooling, snapshots, checksums, compression and other features.
```
Comment (stgraber) [inline doc/reference/storage_btrfs.md:8]: Ah, you actually have RAID mentioned here which definitely contradicts the single-disk part.

File: doc/reference/storage_btrfs.md [hunk 4]
```diff
@@ -1,40 +1,66 @@
...
+This means that users can trivially escape any quotas that are set.
+Therefore, if strict quotas are needed, you should consider using a different storage driver (for example, ZFS with refquotas).
+
+When using quotas, you must take into account that Btrfs extents are immutable.
+When blocks are written, they end up in new extents.
+The old extents remain until all their data is de-referenced or rewritten.
+This means that a quota can be reached even if the total amount of space used by the current files in the subvolume is smaller than the quota.
+
+This issue is seen most often when using VMs on Btrfs, due to the random I/O nature of using raw disk image files on top of a Btrfs subvolume.
+Therefore, you should not use VMs with Btrfs storage pools.
```
Comment (tomponline) [inline doc/reference/storage_btrfs.md:39]: Can we emphasise this somehow?

Comment (tomponline) [reply doc/reference/storage_btrfs.md:39]: Perhaps we can add this whole VM part in a sub-section so we can link to it from the forums (when it inevitably comes up).

File: doc/reference/storage_btrfs.md [hunk 5]
```diff
@@ -1,40 +1,66 @@
...
+If you need to use VMs with Btrfs storage pools anyway, set the instance root disk's {ref}`size.state <instance_device_type_disk>` property to twice the size of the root disk's size.
```
Comment (tomponline) [inline doc/reference/storage_btrfs.md:40]: Suggest "If you really need..."  given that we said don't use VMs with Btrfs the line above.

File: doc/reference/storage_ceph.md [hunk 6]
```diff
@@ -1,56 +1,110 @@
...
-
-## Storage pool configuration
+# Ceph RBD - `ceph`
+
+<!-- Include start Ceph intro -->
+
+```{youtube} https://youtube.com/watch?v=kVLGbvRU98A
+```
+
+[Ceph](https://ceph.io/en/) is an open-source storage platform that stores its data in a storage cluster based on {abbr}`RADOS (Reliable Autonomic Distributed Object Store)`.
```
Comment (tomponline) [inline doc/reference/storage_ceph.md:9]: Maybe just https://ceph.io/ here in case they change their URL formats in the future?

File: doc/reference/storage_ceph.md [hunk 7]
```diff
@@ -1,56 +1,110 @@
...
+They are also referred to as *data pools*, *storage pools* or *OSD pools*.
+<!-- Include end Ceph terminology -->
+
+Ceph block devices are also called *RBD images*, and you can create *snapshots* and *clones* of these RBD images.
+
+## `ceph` driver in LXD
+
+```{note}
+To use the Ceph RBD driver, you must specify it as `ceph`.
+This is slightly misleading, because it provides only Ceph RBD (block storage) functionality, not full Ceph functionality.
```
Comment (tomponline) [inline doc/reference/storage_ceph.md:33]: "Therefore, the `ceph` driver can only be used for storage volumes with content type `block`." - This isn't correct.

While true that the `ceph` storage pool type only uses ceph's underlying block storage functionality, LXD will use ceph RBD block volumes to provide LXD filesystem volumes (for images, containers and custom filesystem volumes), as well as LXD block volumes (for VM images, VMs and custom block volumes).

Instead I propose we change it to something like:

> The `ceph` pool type only uses Ceph RBD (block storage) functionality, not full Ceph functionality.
LXD will create RBD volumes for both filesystem and block LXD volumes.
> It is also possible to use CephFS functionality for custom filesystem volumes by creating a {ref}`storage-cephfs` type pool.

Comment (tomponline) [reply doc/reference/storage_ceph.md:33]: So `ceph` pool type actually supports all LXD functionality, whereas `cephfs` pool type only supports LXD custom filesystem volumes.

File: doc/reference/storage_ceph.md [hunk 8]
```diff
@@ -1,56 +1,110 @@
...
+
+Due to the way copy-on-write works in Ceph RBD, parent RBD images can't be removed until all children are gone.
+As a result, LXD automatically renames any objects that are removed but still referenced.
+Such objects are kept with a  `zombie_` prefix until all references are gone and the object can safely be removed.
+
+### Limitations
+
+The `ceph` driver has the following limitations:
+
+Sharing the OSD storage pool between instances
```
Comment (tomponline) [inline doc/reference/storage_ceph.md:63]: We need to be careful around the term "instances" here, as everywhere else in the docs an "instance" is either a container or a VM, but here we are talking about a LXD installation (I think, given the line below). So perhaps "LXD installation" rather than "LXD instance" would be clearer.

File: doc/reference/storage_ceph.md [hunk 9]
```diff
@@ -1,56 +1,110 @@
...
+: Sharing the same OSD storage pool between multiple LXD instances is not supported.
+  The only scenario in which LXD allows sharing an OSD storage pool between multiple LXD instances is if you want to back up existing instances via `lxd import`.
```
Comment (tomponline) [inline doc/reference/storage_ceph.md:65]: `lxd import` doesn't exist anymore, its been replaced with `lxd recover`, and that is (as the name suggests) not for backup but for recovering a LXD installation after DB corruption. So I'm not sure this section is relevant anymore, but at most it should be reworked to describe how it can be used to recover a LXD installation after DB corruption/loss when there are existing instances on the ceph pool.

File: doc/reference/storage_ceph.md [hunk 10]
```diff
@@ -1,56 +1,110 @@
...
+  In this case, you must set the {ref}`ceph.osd.force_reuse <storage-ceph-pool-config>` configuration option to `true`.
+  If this option is not set, LXD refuses to reuse an OSD storage pool that it detects as being in use by another LXD instance.
```
Comment (tomponline) [inline doc/reference/storage_ceph.md:67]: "LXD installation"

File: doc/reference/storage_ceph.md [hunk 11]
```diff
@@ -1,56 +1,110 @@
...
+Therefore, you should never maintain any file system entities that are not owned by LXD in a LXD OSD storage pool, because LXD might delete them.
+<!-- Include end Ceph driver control -->
+
+Due to the way copy-on-write works in Ceph RBD, parent RBD images can't be removed until all children are gone.
+As a result, LXD automatically renames any objects that are removed but still referenced.
+Such objects are kept with a  `zombie_` prefix until all references are gone and the object can safely be removed.
+
+### Limitations
+
+The `ceph` driver has the following limitations:
```
Comment (tomponline) [inline doc/reference/storage_ceph.md:61]: I think we should also add a note to this section to indicate that a custom volume can only be assigned to a single instance at a time, it does not allow sharing a custom volume between multiple instances on multiple cluster members at the same time (this is what cephfs custom filesystem volumes do allow).

File: doc/reference/storage_ceph.md [hunk 12]
```diff
@@ -1,56 +1,110 @@
...
+
+Sharing the OSD storage pool between instances
+: Sharing the same OSD storage pool between multiple LXD instances is not supported.
+  The only scenario in which LXD allows sharing an OSD storage pool between multiple LXD instances is if you want to back up existing instances via `lxd import`.
+  In this case, you must set the {ref}`ceph.osd.force_reuse <storage-ceph-pool-config>` configuration option to `true`.
+  If this option is not set, LXD refuses to reuse an OSD storage pool that it detects as being in use by another LXD instance.
+
+Choosing a file system
+: When setting up the Ceph cluster that LXD will use, choose XFS as the underlying file system for the storage entities that are used to hold OSD storage pools.
```
Comment (tomponline) [inline doc/reference/storage_ceph.md:70]: Do we have a source reference we can link to as to where this advice comes from?

File: doc/reference/storage_ceph.md [hunk 13]
```diff
@@ -1,56 +1,110 @@
...
+
+The `ceph` driver has the following limitations:
+
+Sharing the OSD storage pool between instances
+: Sharing the same OSD storage pool between multiple LXD instances is not supported.
+  The only scenario in which LXD allows sharing an OSD storage pool between multiple LXD instances is if you want to back up existing instances via `lxd import`.
+  In this case, you must set the {ref}`ceph.osd.force_reuse <storage-ceph-pool-config>` configuration option to `true`.
+  If this option is not set, LXD refuses to reuse an OSD storage pool that it detects as being in use by another LXD instance.
+
+Choosing a file system
```
Comment (tomponline) [inline doc/reference/storage_ceph.md:69]: > Choosing an underlying filesystem 

To avoid confusion with choosing a filesystem for the LXD volumes themselves (controlled by the `volume.block.filesystem` setting).

File: doc/reference/storage_dir.md [hunk 14]
```diff
@@ -1,25 +1,38 @@
...
 # Directory - `dir`
 
- - While this backend is fully functional, it's also much slower than
-   all the others due to it having to unpack images or do instant copies of
-   instances, snapshots and images.
- - Quotas are supported with the directory backend when running on
-   either ext4 or XFS with project quotas enabled at the filesystem level.
+The directory storage driver is a basic backend that stores its data in a standard file and directory structure.
+This driver is quick to set up and allows inspecting the files directly on the disk, which can be convenient for testing.
+However, LXD operations are not optimized for this driver.
```
Comment (tomponline) [inline doc/reference/storage_dir.md:6]: A link to https://linuxcontainers.org/lxd/docs/master/reference/storage_drivers/#feature-comparison seems appropriate here to allow the reader to understand what "not optimized" means.

Comment (tomponline) [reply doc/reference/storage_dir.md:6]: And/or we move the next paragraph up to this section, as that explains it somewhat.

File: doc/reference/storage_dir.md [hunk 15]
```diff
@@ -1,25 +1,38 @@
...
 
-## Storage pool configuration
+## `dir` driver in LXD
+
+The `dir` driver in LXD is fully functional and provides the same set of features as other drivers.
+However, it is much slower than all the other drivers because it must unpack images and do instant copies of instances, snapshots and images.
+
+Unless specified differently during creation (with the `source` configuration option), the data is stored in the `/var/snap/lxd/common/lxd/storage-pools/` directory.
+
+### Quotas
```
Comment (tomponline) [inline doc/reference/storage_dir.md:15]: Should the link on the tiny "1" at https://linuxcontainers.org/lxd/docs/master/reference/storage_drivers/#note link to here now?

File: doc/reference/storage_lvm.md [hunk 16]
```diff
@@ -1,45 +1,62 @@
...
+It is used to manage physical storage devices, allowing you to create a number of logical storage volumes that use and virtualize the underlying physical storage devices.
+
+Note that it is possible to over-commit the physical storage in the process, to allow flexibility for scenarios where not all available storage is in use at the same time.
+
+## Terminology
+
+LVM can combine several physical storage devices into a *volume group*.
+You can then allocate *logical volumes* of different types from this volume group.
+
+One supported volume type is a *thin pool*, which allows over-committing the resources by creating volumes that are larger than the available physical storage.
```
Comment (tomponline) [inline doc/reference/storage_lvm.md:14]: "One supported volume type is a *thin pool*, which allows over-committing the resources by creating thinly provisioned volumes whose total allocated size is larger than the available physical storage."

File: doc/reference/storage_lvm.md [hunk 17]
```diff
@@ -1,45 +1,62 @@
...
+Another type is a *volume snapshot*, which captures a specific state of a logical volume.
```
Comment (tomponline) [inline doc/reference/storage_lvm.md:15]: I'm not sure if you want to go into it at this stage, but volume snapshots of thin vs non-thin logical volumes behave quite differently. 

For snapshots of thin logical volumes, you don't have to specify a size of the snapshot, and they don't consume allocated space until writes happen on the source logical volume. 

However with non-thin snapshots one has to specify a maximum size of the snapshot at creation time, this size is then taken away from the available space in the volume group. Note, this size is used to allow the volume snapshot to accommodate recording the differences as writes made to the source volume. 

This means that if a snapshot size is too small, then at some point in the future, depending on how many changes are made to the source volume, the snapshot can become "invalidated" and no longer usable. As you can image, as snapshots can be used as a form of "backup", this sort of behaviour is undesirable. So when LXD creates non-thin snapshots it uses the maximum size in order to accommodate the entire source volume from being re-written. This turns out to be essential with VMs because LVM non-thin doesn't support restoring a snapshot and leaving it the snapshot intact (unlike LVM thin volumes, BTRFS and ZFS). So in order to not destroy the snapshot when it is restored, we actually have to manually copy the contents of the snapshot over the top of the original volume (i.e we perform writes to the entire source volume). If we were not to set the snapshot size to the maximum size then restoring a snaphot would invalidate all the other snapshots (including itself).

So long story short, for LVM non-thin snaphots, they effectively take up the space of the source volume at create time.

See https://github.com/lxc/lxd/blob/master/lxd/storage/drivers/driver_lvm_volumes.go#L1203-L1224

File: doc/reference/storage_lvm.md [hunk 18]
```diff
@@ -1,45 +1,62 @@
...
+
+## `lvm` driver in LXD
+
+The `lvm` driver in LXD uses logical volumes for images, and volume snapshots for instances and snapshots.
+The file system used for the logical volumes is ext4 by default.
+You can configure the `lvm` driver to use XFS instead by setting the {ref}`block.filesystem <storage-lvm-vol-config>` configuration for the volume (or the corresponding `volume.block.filesystem` configuration on the storage pool for all volumes in the pool).
```
Comment (tomponline) [inline doc/reference/storage_lvm.md:21]: It feels like we should have a generic section about what values are supported by the `block.filesystem` (and by association the `volume.block.filesystem`) setting for all supported storage pools, rather than duplicate that. 

Because they are the same independent of storage pool type (for those storage pools that support the `block.filesystem` setting), they are: btrfs, ext4 and xfs. Defaulting to ext4.

File: doc/reference/storage_lvm.md [hunk 19]
```diff
@@ -1,45 +1,62 @@
...
+
+By default, LVM storage pools use an LVM thin pool and create logical volumes for all LXD storage entities (images, instances etc.) in there.
+This behavior can be changed by setting {ref}`lvm.use_thinpool <storage-lvm-pool-config>` to `false`.
```
Comment (tomponline) [inline doc/reference/storage_lvm.md:24]: "This behaviour can be changed at pool create time..."

It cannot be changed later.

File: doc/reference/storage_lvm.md (hunk 17 above)
Comment (tomponline) [reply doc/reference/storage_lvm.md:15]: Ah I see you've covered some of this here https://github.com/lxc/lxd/pull/10643/files#diff-07ed16fddda9ff1e8130b444e649bd1086ece89e46c3c353ff66270dccc97271R25-R28

Maybe just add a mention below that as well as performance reductions, there is also space reduction.

File: doc/reference/storage_zfs.md [hunk 20]
```diff
@@ -5,70 +5,111 @@ discourse: 1333
...
+These datasets can be of different types:
+- A *ZFS filesystem* can be seen as a partition or a mounted file system.
+- A *ZFS volume* represents a block device.
+- A *ZFS snapshot* captures a specific state of either a ZFS filesystem or a ZFS volume.
+  ZFS snapshots are read-only.
+- A *ZFS clone* is a writable copy of a ZFS snapshot.
+
+## `zfs` driver in LXD
+
+The `zfs` driver in LXD uses ZFS filesystems for images, and ZFS snapshots and clones to create instances and snapshots.
```
Comment (tomponline) [inline doc/reference/storage_zfs.md:30]: "The `zfs` driver in LXD uses ZFS filesystems and volumes for images and custom volumes, and ZFS snapshots and clones to create instances from images and for instance and custom volume snapshots."

File: doc/reference/storage_zfs.md [hunk 21]
```diff
@@ -5,70 +5,111 @@ discourse: 1333
...
+By default, LXD enables compression when creating a ZFS pool.
+
+LXD assumes that it has full control over the ZFS pool and dataset.
```
Comment (tomponline) [inline doc/reference/storage_zfs.md:33]: Can we add a similar statement to the LVM section, but rather than ZFS pool & dataset, it would be volume group for LVM.

File: doc/reference/storage_zfs.md [hunk 22]
```diff
@@ -5,70 +5,111 @@ discourse: 1333
...
+Therefore, you should never maintain any datasets or file system entities that are not owned by LXD in a LXD ZFS pool or dataset, because LXD might delete them.
+
+Due to the way copy-on-write works in ZFS, parent ZFS filesystems can't be removed until all children are gone.
+As a result, LXD automatically renames any objects that are removed but still referenced.
+Such objects are kept at a random `deleted/` path until all references are gone and the object can safely be removed.
```
Comment (tomponline) [inline doc/reference/storage_zfs.md:38]: Maybe add "This also has ramifications for restoring snapshots see Limitations"

**Review by tomponline — CHANGES_REQUESTED** [review]
A few comments/suggestions.

I also think we would benefit from having a link to https://linuxcontainers.org/lxd/docs/master/howto/storage_configure/#configure-default-values-for-storage-volumes in each storage pool type's config settings section with a mention that the `volume.*` settings can be specified, as they aren't listed in the pool config options table.

File: doc/reference/storage_btrfs.md (hunk 4 above)
Comment (ru-fu) [reply doc/reference/storage_btrfs.md:39]: I put it in a note now, which should emphasise it. I don't think it would work as a sub-section, since the explanation is in the Quotas section, so you'd need to link there anyway.

File: doc/reference/storage_ceph.md (hunk 9 above)
Comment (ru-fu) [reply doc/reference/storage_ceph.md:65]: Who would know if it is still relevant?
And any info on what else we need here (apart from the `lxd recover` help, which I don't think we need to replicate here)?

File: doc/reference/storage_dir.md (hunk 15 above)
Comment (ru-fu) [reply doc/reference/storage_dir.md:15]: Good catch!

File: doc/reference/storage_ceph.md (hunk 12 above)
Comment (tomponline) [reply doc/reference/storage_ceph.md:70]: Maybe @stgraber knows why?

File: doc/reference/storage_ceph.md (hunk 9 above)
Comment (tomponline) [reply doc/reference/storage_ceph.md:65]: Any thoughts @stgraber ?

File: doc/reference/storage_lvm.md (hunk 18 above)
Comment (ru-fu) [reply doc/reference/storage_lvm.md:21]: The option is only listed for LVM and Ceph RBD, or is it missing somewhere?
If it's only those two, I think it's fine to mention it in both places (since there is a limitation for XFS for Ceph RBD). But I'll also add the options and default to the table.

Comment (tomponline) [reply doc/reference/storage_lvm.md:21]: Yes its only relevant for LVM and Ceph. I'm not sure what you mean about the XFS limitation for Ceph RBD, here we are talking about the filesystems used on the Ceph RBD (block volumes) rather than the filesystem used to host Ceph itself. I've checked the code and both LVM and Ceph drivers allow the same set of filesystems.

Comment (ru-fu) [reply doc/reference/storage_lvm.md:21]: Ah okay - yes, I confused those two. Then we can take it out here as well.

File: doc/reference/storage_zfs.md [hunk 23]
```diff
@@ -5,70 +5,111 @@ discourse: 1333
...
+  This method makes it possible to confirm whether a specific snapshot contains what you need.
+  After determining the correct snapshot, you can {ref}`remove the newer snapshots <storage-edit-snapshots>` so that the snapshot you need is the latest one and you can restore it.
+
+  Alternatively, you can configure LXD to automatically discard the newer snapshots during restore.
+  To do so, set the {ref}`zfs.remove_snapshots <storage-zfs-vol-config>` configuration for the volume (or the corresponding `volume.zfs.remove_snapshots` configuration on the storage pool for all volumes in the pool).
+
+  Note, however, that instance copies use ZFS snapshots too.
+  Therefore, you cannot restore an instance to a snapshot taken before the last copy without having to also delete all its descendants.
+  If this is not an option, you can copy the wanted snapshot into a new instance and then delete the old instance.
+  You will, however, lose any other snapshots the instance might have had.
```
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/reference/storage_zfs.md:69]: Does that mean that `zfs.clone_copy` must be set to "false" or "rebase" so that the copying workaround will work?
Or that this limitation only applies if `zfs.clone_copy` is set to "false" or "rebase"?

File: doc/reference/storage_zfs.md [hunk 24]
```diff
@@ -5,70 +5,111 @@ discourse: 1333
...
+
+Observing I/O quotas
+: I/O quotas (IOps/MBs) are unlikely to affect ZFS filesystems very much.
```
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/reference/storage_zfs.md:72]: Google says IOPS, actually.

[the remark this answers is not in the document]
Comment (tomponline) [reply doc/reference/storage_zfs.md]: Lol, in that case it should also be MBS to be consistent (the s in both being "per second"), but that is normally written as MBs or MB/s. Shame we can't have consistency there. Id be tempted to remove the part in brackets entirely.

**Comment by ru-fu** [comment]
Thanks for the comments! Updated. :)

File: doc/reference/storage_zfs.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (tomponline) [reply doc/reference/storage_zfs.md:72]: This was really supposed to be referencing the line above that says "Note, however, that instance copies use ZFS snapshots too.", this is true if `zfs.clone_copy` is set to the default `true`. However if set to `false` or `rebase` then instance copies are done by duplicating the volume (either fully in the case of `false` or in the case of `rebase` then using the source instance's original image snapshot as the basis of the copy).

In those circumstances, instance copies are not snapshots of the source instance, and thus they do not contribute toward any restriction with regard to restoring snapshots of the source instance.

File: doc/reference/storage_zfs.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/reference/storage_zfs.md]: I guess it isn't consistent anyway and should be MBPS then. ;) But let's just take it out.

File: doc/reference/storage_zfs.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/reference/storage_zfs.md:72]: Thanks! Fixed.

File: doc/reference/storage_zfs.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (tomponline) [reply doc/reference/storage_zfs.md]: Yeah agreed take it out. Just for more confusion megabits per second is normally written as "Mbps" (lowercase b), so Megabytes per second can be "MB/s" or "MBps" (https://en.wikipedia.org/wiki/Data-rate_units).

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
Thanks!
